python/src/equistore/operations/unique_metadata.py

```diff
@@ -0,0 +1,245 @@
+
+    If there are no indices in the ``block`` (or its gradient) corresponding to
+    the specified ``axis`` and ``names``, an empty Labels object with the
+    correct names as in the passed ``names`` is returned.
+
+    For example, to find the unique "structure" indices in the "samples"
+    metadata present in a given TensorBlock:
+
+    .. code-block:: python
+
+        unique_samples = unique_metadata_block(
+            block, axis="samples", names=["structure"],
+        )
+
+    To find the unique "atom" indices along the "samples" axis present in the
+    "positions" gradient block of a given TensorBlock:
+
+    .. code-block:: python
+
+        unique_grad_samples = unique_metadata_block(
+            block, axis="samples", names=["atom"], gradient_param="positions",
+        )
+
+    :param block: the :py:class:`TensorBlock` to find unique indices for.
+    :param axis: a str, either "samples" or "properties", corresponding to the
+        axis along which the named unique metadata should be found.
+    :param names: a str, list of str, or tuple of str corresponding to the
+        name(s) of the metadata along the specified ``axis`` for which the
+        unique indices should be found.
+    :param gradient_param: a str corresponding to the gradient parameter name
+        for the gradient blocks to find the unique metadata for. If none
+        (default), the unique metadata of the regular TensorBlocks will be
+        calculated.
+
+    :return: a sorted :py:class:`Labels` object containing the unique metadata
+        for the input ``block`` or its gradient for the specified parameter.
+        Each element in the returned :py:class:`Labels` object has len(names)
+        entries.
+    """
+    # Parse input args
+    if not isinstance(block, TensorBlock):
+        raise TypeError("``block`` must be an equistore TensorBlock")
+    names = (
+        [names]
+        if isinstance(names, str)
+        else (list(names) if isinstance(names, tuple) else names)
+    )
+    _check_args(block, axis, names, gradient_param)
+    # Make a list of the blocks to find unique indices for
+    if gradient_param is None:
+        blocks = [block]
+    else:
+        blocks = [block.gradient(gradient_param)]
+
+    return _unique_from_blocks(blocks, axis, names, gradient_param)
+
+
+def _unique_from_blocks(
+    blocks: List[TensorBlock],
+    axis: str,
+    names: List[str],
+    gradient_param: Optional[str],
+) -> Labels:
+    """
+    Finds the unique metadata of a list of blocks along the given ``axis`` and
+    for the specified ``names``. If ``gradient_param`` is specified, only finds
+    the unique indices for gradient blocks under the specified parameter name.
+    """
+    # Extract indices from each block
+    all_idxs = []
+    for block in blocks:
+        idxs = block.samples[names] if axis == "samples" else block.properties[names]
+        for idx in idxs:
+            all_idxs.append(idx)
```

Comment:
Maybe something like?

```suggestion
        all_idxs += idxs.tolist()
```